### ISSUE TITLE
[Bug] PE version PDF Export fails?

### ISSUE THREAD

**Issue by KenSense** [thingsboard/thingsboard#5960]
**Describe the bug**

**Your Server Environment**
* own setup
  * local infrastructure
  * ThingsBoard 3.3.2
  * OS Ubuntu 20.04 LTS

**Your Client Environment**
* OS:  Windows 10
* Browser Edge Version 97.0.1072.69 (Official build) (64-bit)

**To Reproduce**
1. Go to 'Dashboard'
2. Click on 'Export Dashboard'
3. Scroll down to 'Export as PDF'

**Expected behavior**
I expect to get a PDF file.
But I get:
{"status":503,"message":null,"errorCode":2,"timestamp":"2022-01-25T11:17:20.277+00:00","subscriptionErrorCode":null,"subscriptionEntry":null,"subscriptionValue":null}

**Screenshots**
<img width="783" alt="2022-01-25_12-17-36" src="https://user-images.githubusercontent.com/49442805/150969645-d77905b7-a0f1-44b4-9a3b-da38b8e69a90.png">
**Additional context**
I have not got it to work yet for the first time, so maybe some configuration settings need to be changed after installing https://dist.thingsboard.io/tb-web-report-3.3.2pe.deb

**Comment by volodymyr-babak** [comment]
hi @KenSense 

please attach logs of Thinsgboard and TB report services. 
Usually, they are located in
/var/log/thingsboard/thingsboard.log
/var/log/tb-web-report/tb-web-report.log

**Comment by KenSense** [comment]
This is from tb-web-report.log 
`2022-01-25 11:16:50,111 [ReportController] INFO: Generating dashboard report: http://ourcompany.com:**80**/dashboard/91e294b0-37d5-11ec-b3e7-49190e746d45
2022-01-25 11:18:50,332 [ReportController] ERROR: Navigation timeout of 120000 ms exceeded:
TimeoutError: Navigation timeout of 120000 ms exceeded
    at /snapshot/web-report/node_modules/puppeteer/lib/cjs/puppeteer/common/LifecycleWatcher.js:106:111
    at async FrameManager.navigateFrame (/snapshot/web-report/node_modules/puppeteer/lib/cjs/puppeteer/common/FrameManager.js:88:21)
    at async Frame.goto (/snapshot/web-report/node_modules/puppeteer/lib/cjs/puppeteer/common/FrameManager.js:405:16)
    at async Page.goto (/snapshot/web-report/node_modules/puppeteer/lib/cjs/puppeteer/common/Page.js:826:16)
    at async generateDashboardReport (/snapshot/web-report/api/controllers/tbWebReportController.js)
`

I see port 80? that is probably a problem, internally thingsboard is behind haproxy on port 8443
Where can I change that so the report generater knows where to find the thingsboard?

In thingsboard.log I didn't see anything but it keeps filling up with this kind of log is that normal?
`2022-01-25 16:01:24,053 [TB-Scheduling-1] INFO  o.t.s.s.s.DefaultTbEntityDataSubscriptionService - Stats: regularQueryInvocationCnt = [0], regularQueryInvocationTime = [0], dynamicQueryCnt = [3] dynamicQueryInvocationCnt = [0], dynamicQueryInvocationTime = [0], alarmQueryInvocationCnt = [0], alarmQueryInvocationTime = [0]
2022-01-25 16:01:26,566 [sql-log-1-thread-1] INFO  o.t.s.dao.sql.TbSqlBlockingQueue - Queue-1 [Attributes] queueSize [0] totalAdded [2] totalSaved [2] totalFailed [0]
2022-01-25 16:01:27,142 [sql-log-1-thread-1] INFO  o.t.s.dao.sql.TbSqlBlockingQueue - Queue-1 [TS] queueSize [0] totalAdded [24] totalSaved [24] totalFailed [0]
2022-01-25 16:01:27,207 [sql-log-1-thread-1] INFO  o.t.s.dao.sql.TbSqlBlockingQueue - Queue-1 [TS Latest] queueSize [0] totalAdded [24] totalSaved [24] totalFailed [0]
2022-01-25 16:01:28,851 [queue-scheduler-15-thread-1] INFO  o.t.s.q.u.DefaultTbApiUsageReportClient - Reporting API usage statistics for 2 tenants and customers
2022-01-25 16:01:34,053 [TB-Scheduling-1] INFO  o.t.s.s.s.DefaultTbEntityDataSubscriptionService - Stats: regularQueryInvocationCnt = [0], regularQueryInvocationTime = [0], dynamicQueryCnt = [3] dynamicQueryInvocationCnt = [0], dynamicQueryInvocationTime = [0], alarmQueryInvocationCnt = [0], alarmQueryInvocationTime = [0]
2022-01-25 16:01:36,566 [sql-log-1-thread-1] INFO  o.t.s.dao.sql.TbSqlBlockingQueue - Queue-1 [Attributes] queueSize [0] totalAdded [2] totalSaved [2] totalFailed [0]
2022-01-25 16:01:36,566 [sql-log-1-thread-1] INFO  o.t.s.dao.sql.TbSqlBlockingQueue - Queue-2 [Attributes] queueSize [0] totalAdded [1] totalSaved [1] totalFailed [0]
2022-01-25 16:01:36,567 [sql-log-1-thread-1] INFO  o.t.s.dao.sql.TbSqlBlockingQueue - Queue-3 [Attributes] queueSize [0] totalAdded [1] totalSaved [1] totalFailed [0]
2022-01-25 16:01:37,142 [sql-log-1-thread-1] INFO  o.t.s.dao.sql.TbSqlBlockingQueue - Queue-1 [TS] queueSize [0] totalAdded [28] totalSaved [28] totalFailed [0]
2022-01-25 16:01:37,206 [sql-log-1-thread-1] INFO  o.t.s.dao.sql.TbSqlBlockingQueue - Queue-1 [TS Latest] queueSize [0] totalAdded [28] totalSaved [28] totalFailed [0]
2022-01-25 16:01:38,851 [queue-scheduler-15-thread-1] INFO  o.t.s.q.u.DefaultTbApiUsageReportClient - Reporting API usage statistics for 2 tenants and customers
2022-01-25 16:01:39,499 [tb-rule-engine-consumer-43-thread-27 | Main-2] INFO  o.t.s.s.q.DefaultTbRuleEngineConsumerService - Failed to process [2] messages`

**Comment by Backdraft007** [comment]
Do you have upgraded Thingsboard?
If yes, do you have installed latest version of TB-Web-Report?

**Comment by KenSense** [comment]
> Do you have upgraded Thingsboard? If yes, do you have installed latest version of TB-Web-Report?

Web report version 3.3.2 is the first I installed.
But as I said it's probably the report generator that doesn't know on which port to find the thingsboard service since it shows port 80 in the log, that isn't used, I just don't know where I have to set that.
Since there is a log means that Thingsboard can give the command to the generator it just doesn't know where to find the dashboard to convert it to a PDF.

**Comment by Backdraft007** [comment]
Can you post your haproxy.cfg?

**Comment by Backdraft007** [comment]
I'm not an expert with haproxy. But I think bind :80 is there not optimal. Can you try to remove the bind :80?

This config works for me: https://thingsboard.io/docs/user-guide/install/pe/add-haproxy-ubuntu/#step-5-configure-haproxy-load-balancer

And in thingsboard.yml is the server port also 8443?

**Comment by KenSense** [comment]
I don't think haproxy has anything to do with it, since none of the ports that the report generator needs is handled by haproxy.

**Comment by Backdraft007** [comment]
Are you using https in your browser?

When I export a dashboard, I get this:
[ReportController] INFO: Generating dashboard report: https://domain:443/dashboard/b44bae60-24fc-11ec-9c54-7f39658c0a89
[ReportController] INFO: Report data sent.

I have checked all configurations and found nothing, so I think TB take the url from the client browser.

**Comment by KenSense** [comment]
> Are you using https in your browser?
Yes that is mainly why haproxy is present, even if I enter :80 it will go to 443 https (but could be my browser doing that instead of haproxy)

Is your thingsboard.log file also getting so big? (see my first reply)

**Comment by volodymyr-babak** [comment]
@KenSense 
I suppose @Backdraft007 is correct in his assumptions. 
TB web report service doesn't aware of the TB service behind a proxy, TB web report is using URL from the 'client'.
Here is the code snippet how port is selected:
`    public static int getPort(HttpServletRequest request){
        String forwardedProto = request.getHeader("x-forwarded-proto");

        int serverPort = request.getServerPort();
        if (request.getHeader("x-forwarded-port") != null) {
            try {
                serverPort = request.getIntHeader("x-forwarded-port");
            } catch (NumberFormatException e) {
            }
        } else if (forwardedProto != null) {
            switch (forwardedProto) {
                case "http":
                    serverPort = 80;
                    break;
                case "https":
                    serverPort = 443;
                    break;
            }
        }
        return serverPort;
    }`

And here is the snippet of how schema is selected:
`public static String getScheme(HttpServletRequest request){
        String scheme = request.getScheme();
        String forwardedProto = request.getHeader("x-forwarded-proto");
        if (forwardedProto != null) {
            scheme = forwardedProto;
        }
        return scheme;
    }`

I see that in your configuration you are using TLS on HAProxy and TB as well and I suspect, that this is the problem.
You'll need to make sure that correct headers are provided in the request.
So probably in your case you'll need to add **x-forwarded-proto** as **https** and **x-forwarded-port** as **443** to make sure that web report component get's a valid URI to generate the report.

**Comment by KenSense** [comment]
It's only doing TLS in haproxy not in thingsboard.
But your second point is correct it's looking for header info that is not present.
I added the needed lines in haproxy and now it is looking for 443 the link in the error log works BUT it still doesn't generate a PDF.

So any ideas?

I get this timeout error again
{"status":503,"message":null,"errorCode":2,"timestamp":"2022-01-```
31T08:27:14.560+00:00","subscriptionErrorCode":null,"subscriptionEntry":null,"subscriptionValue":null}
```
2022-01-31 08:23:40,294 [main] INFO: Bind address: 0.0.0.0
2022-01-31 08:23:40,295 [main] INFO: Bind port: 8383
2022-01-31 08:23:40,295 [main] INFO: Starting chrome headless browser...
2022-01-31 08:23:40,295 [main] INFO: Chrome headless browser executable: /usr/share/tb-web-report/chromium/chrome
2022-01-31 08:23:41,595 [main] INFO: Started chrome headless browser.
2022-01-31 08:23:41,597 [main] INFO: Headless chrome browser version: HeadlessChrome/85.0.4182.0
2022-01-31 08:23:41,600 [main] INFO: ==> 🌎  ThingsBoard Web Reporting Service listening on http://0.0.0.0:8383/.
2022-01-31 08:26:43,912 [ReportController] INFO: Generating dashboard report: https://mycompany.com:443/dashboard/df20d150-c544-11eb-ae20-b93d26b8058b
2022-01-31 08:28:44,075 [ReportController] ERROR: Navigation timeout of 120000 ms exceeded:
TimeoutError: Navigation timeout of 120000 ms exceeded
    at /snapshot/web-report/node_modules/puppeteer/lib/cjs/puppeteer/common/LifecycleWatcher.js:106:111
    at async FrameManager.navigateFrame (/snapshot/web-report/node_modules/puppeteer/lib/cjs/puppeteer/common/FrameManager.js:88:21)
    at async Frame.goto (/snapshot/web-report/node_modules/puppeteer/lib/cjs/puppeteer/common/FrameManager.js:405:16)
    at async Page.goto (/snapshot/web-report/node_modules/puppeteer/lib/cjs/puppeteer/common/Page.js:826:16)
    at async generateDashboardReport (/snapshot/web-report/api/controllers/tbWebReportController.js)
```

Updated haproxy.cfg:
```
frontend mycompany.com
    mode http
    bind :80
    bind :443 ssl crt /etc/ssl/certs/Cert_2021-06-10.pem
    http-request redirect scheme https code 301 unless { ssl_fc }
    http-response set-header Strict-Transport-Security "max-age=16000000;"
    option forwardfor
    http-request set-header X-Forwarded-Proto https
    default_backend thingsboardservers

backend thingsboardservers
  http-request set-header X-Forwarded-Port 443
  server server1 127.0.0.1:8443

**Comment by volodymyr-babak** [comment]
@KenSense 
is URL https://mycompany.com:443/dashboard/df20d150-c544-11eb-ae20-b93d26b8058b accessible by web-report container?

**Comment by KenSense** [comment]
> @KenSense is URL https://mycompany.com:443/dashboard/df20d150-c544-11eb-ae20-b93d26b8058b accessible by web-report container?

It seems lynx is not working to our public websites from that sever (also time out)
The websites are working without problem from outside.
lynx to other public sides works, I don't know....
Must be some special firewall setting outside the VM....(since it has to go outside to public IP to go back to inside IP server)

**Comment by medutySMST** [comment]
Hello I am having the same issue on docker deployment, 

I am using traefik as a router and ssl handler. However as I understood this feed, the proxy did not seem to be the solution ultimately ? 

error is also `{"status":503,"message":null,"errorCode":2,"timestamp":"2022-04-13T13:02:04.463+00:00","subscriptionErrorCode":null,"subscriptionEntry":null,"subscriptionValue":null} `

my tb-web-report-log says: 

```
2022-05-13 06:15:20,860 [main] INFO: Bind address: 0.0.0.0
2022-05-13 06:15:20,861 [main] INFO: Bind port: 8383
2022-05-13 06:15:20,861 [main] INFO: Starting chrome headless browser...
2022-05-13 06:15:20,862 [main] INFO: Chrome headless browser executable: /usr/share/tb-web-report/chromium/chrome
2022-05-13 06:15:21,080 [main] INFO: Started chrome headless browser.
2022-05-13 06:15:21,081 [main] INFO: Headless chrome browser version: HeadlessChrome/97.0.4691.0
2022-05-13 06:15:21,086 [main] INFO: ==> 🌎  ThingsBoard Web Reporting Service listening on http://0.0.0.0:8383/.
2022-05-13 06:17:37,958 [ReportController] INFO: Generating dashboard report: https://u2xu5.smart-structures.at:443/dashboard/5a7016f0-b571-11ec-a0e3-bd4858cf551f
2022-05-13 06:19:38,078 [ReportController] ERROR: Navigation timeout of 120000 ms exceeded:
TimeoutError: Navigation timeout of 120000 ms exceeded
    at /snapshot/web-report/node_modules/puppeteer/lib/cjs/puppeteer/common/LifecycleWatcher.js:107:111
    at async FrameManager.navigateFrame (/snapshot/web-report/node_modules/puppeteer/lib/cjs/puppeteer/common/FrameManager.js:130:21)
    at async Frame.goto (/snapshot/web-report/node_modules/puppeteer/lib/cjs/puppeteer/common/FrameManager.js:500:16)
    at async Page.goto (/snapshot/web-report/node_modules/puppeteer/lib/cjs/puppeteer/common/Page.js:1167:16)
    at async generateDashboardReport (/snapshot/web-report/api/controllers/tbWebReportController.js)
```

Could you elaborate on how you got this fixed ?

**Comment by KenSense** [comment]
> Could you elaborate on how you got this fixed ?

My first problem was it wasn't using port :443, that is not a issue in your log.
My second problem was that I could not get outside using the domain name so I added the rule to point the domain name to the internal servers IP. (so it sees the same server as using localhost)
If you don't have this second problem, then you have another issue, I can' t help there...

**Comment by medutySMST** [comment]
I found out that when exporting a report via scheduler and setting localhost as URL the export works, so I guess I have the second problem. 

However I am not sure where to set this host rule. 

So at the moment I have a VIM with tb container and traefik (proxy) container. 

Is this rule something to set on the VIM, traefik or in the tb container ?

**Comment by KenSense** [comment]
I did set it in the Linux VM where the tb is installed on. I didn't use a docker container.

**Comment by medutySMST** [comment]
Then I guess it is something to configure inside of the tb container, since the conatiner mimics a VM with tb installed on it. 

Maybe this workaround will work: 
     customize the entrypoint or dockerfile to add 
     internalserverip mycompany.com 
     to /etc/hosts
as you did it.

However this does feel like a hack since you are not supposed to mess with those things. (entrypoint or dockerfile) (right?)

I will have to take another look, I also got a suggestion from the tb support to set up a seperate tb-web-report container, still looking into that solution. 

I also opened another issue since this is similar but not the same. 
If you have any insight, I would be glad if you drop by the new thread.  ([Dashboard PDF export not working [Bug] #6574](https://github.com/thingsboard/thingsboard/issues/6574))

Best regards
MD
